tracing-appender/src/non_blocking.rs

```diff
@@ -144,3 +144,124 @@ impl Drop for WorkerGuard {
         }
     }
 }
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use rand::Rng;
+    use std::sync::Mutex;
+    use std::thread;
+    use std::time::Duration;
+
+    struct MockWriter {
+        writer: Arc<Mutex<Vec<String>>>,
+        max_writes_allowed: usize,
+        writes_attempted: usize,
+    }
+
+    impl Default for MockWriter {
+        fn default() -> Self {
+            MockWriter {
+                writer: Arc::new(Mutex::new(Vec::new())),
+                max_writes_allowed: 1,
+                writes_attempted: 0,
+            }
+        }
+    }
+
+    impl std::io::Write for MockWriter {
+        fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+            let buf_len = buf.len();
+            self.writes_attempted += 1;
+            if self.writes_attempted > self.max_writes_allowed {
+                return Err(std::io::Error::from(std::io::ErrorKind::WouldBlock));
+            }
+
+            self.writer
+                .lock()
+                .expect("expected guard")
+                .push(String::from_utf8_lossy(buf).to_string());
+            Ok(buf_len)
+        }
+
+        fn flush(&mut self) -> std::io::Result<()> {
+            Ok(())
+        }
+    }
+
+    #[test]
+    fn logs_dropped_if_lossy() {
+        let mock_writer = MockWriter::default();
+
+        let (mut non_blocking, _guard) = self::NonBlockingBuilder::default()
+            .lossy(true)
+            .buffered_lines_limit(1)
+            .finish(mock_writer);
+
+        let error_count = non_blocking.error_counter();
+
+        non_blocking
+            .write_all("Hello".as_bytes())
+            .expect("Failed to write");
+        assert_eq!(0, error_count.load(Ordering::Relaxed));
+
+        non_blocking
+            .write_all(", World".as_bytes())
+            .expect("Failed to write");
+        assert_eq!(1, error_count.load(Ordering::Relaxed));
+
+        non_blocking.write_all(".".as_bytes()).expect("Failed to write");
+        assert_eq!(2, error_count.load(Ordering::Relaxed));
+    }
+
+    #[test]
+    fn multi_threaded_writes() {
+        let inner_writer = Arc::new(Mutex::new(Vec::new()));
+
+        let mut mock_writer = MockWriter::default();
+        mock_writer.max_writes_allowed = DEFAULT_BUFFERED_LINES_LIMIT;
+        mock_writer.writer = inner_writer.clone();
+
+        let (non_blocking, _guard) = self::NonBlockingBuilder::default()
+            .lossy(true)
+            .finish(mock_writer);
+
+        let error_count = non_blocking.error_counter();
+        let mut join_handles: Vec<JoinHandle<()>> = Vec::with_capacity(10);
+
+        let subscriber = tracing_subscriber::fmt().with_writer(non_blocking.clone());
+
+        tracing::subscriber::with_default(subscriber.finish(), || {
```

Comment:
This won't be set as the default subscriber for the spawned threads --- we'd need to call `with_default` _inside_ of the `thread::spawn` closure.

Also, I don't think the subscriber is ever being used, since we are just calling `write` on the non_blocking writer directly?

Reply:
Sorry, this specific test is kind of a mess. I was trying multiple different things and in my haste to get the PR out, I forgot to remove the subscriber since I wasn't using it anymore.